### PR TITLE
Support .netrc files as a way to authenticate HTTP(S) downloads

### PR DESCRIPTION
**Pull request by sschuberth**
This includes support for cloning private Git repos via HTTP(S). Please have a look at the individual commit messages for the details.

### CODE REVIEW

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 1]
```diff
@@ -74,15 +74,54 @@ class OrtAuthenticator(private val original: Authenticator? = null) : Authentica
...
+                    return proxySelector.getProxyAuthentication(proxy)
+                }
+            }
+
+            RequestorType.SERVER -> {
+                val netrcFile = getUserHomeDirectory().resolve(".netrc")
+                if (netrcFile.isFile) {
+                    getNetrcAuthentication(netrcFile.readText(), requestingHost)?.let { return it }
+                }
+
```
Comment (fviernau) [inline utils/src/main/kotlin/OrtAuthenticator.kt:92]: Would it make sense to also log a warning if we reach this line?

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 2]
```diff
@@ -74,15 +74,54 @@ class OrtAuthenticator(private val original: Authenticator? = null) : Authentica
...
+            null -> log.warn { "No requestor type set for password authentication." }
         }
 
         return super.getPasswordAuthentication()
     }
 }
+
+fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentication? {
+    val lines = contents.lines().mapNotNull { line ->
```
Comment (fviernau) [inline utils/src/main/kotlin/OrtAuthenticator.kt:105]: What do you think about adding the URL to the netrc file format here as a code comment next to the parsing?

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 3]
```diff
@@ -74,15 +74,54 @@ class OrtAuthenticator(private val original: Authenticator? = null) : Authentica
     }
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt:74]: Done.

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 4]
```diff
@@ -74,15 +74,54 @@ class OrtAuthenticator(private val original: Authenticator? = null) : Authentica
...
+            }
+
+            RequestorType.SERVER -> {
+                val netrcFile = getUserHomeDirectory().resolve(".netrc")
+                if (netrcFile.isFile) {
+                    getNetrcAuthentication(netrcFile.readText(), requestingHost)?.let { return it }
+                }
+
+                // TODO: Add support for .authinfo files (which use the same syntax as .netrc files) once Git.kt does
+                //       not call the Git CLI anymore which only suuports .netrc files.
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt:94]: Done.

File: utils/src/main/kotlin/OrtAuthenticator.kt [hunk 5]
```diff
@@ -74,15 +74,54 @@ class OrtAuthenticator(private val original: Authenticator? = null) : Authentica
...
+                    val type = requestingProtocol.toProxyType() ?: return null
+                    val proxy = Proxy(type, InetSocketAddress(requestingHost, requestingPort))
+                    return proxySelector.getProxyAuthentication(proxy)
+                }
+            }
+
+            RequestorType.SERVER -> {
+                val netrcFile = getUserHomeDirectory().resolve(".netrc")
+                if (netrcFile.isFile) {
+                    getNetrcAuthentication(netrcFile.readText(), requestingHost)?.let { return it }
```
[the remark this answers is not in the document]
Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt:90]: Note that we should only return if `getNetrcAuthentication()` returns a value `!= null`, i.e. with your suggestion (if I got it right) I'd need to spend another explicit `if`, which is a bit too Java-ish for my taste.

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 1 above)
Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt:92]: I don't think so. If at all maybe an `info`, but I believe even not doing that is fine.

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (fviernau) [reply utils/src/main/kotlin/OrtAuthenticator.kt:90]: Sure, I'm ok with that.

File: utils/src/main/kotlin/OrtAuthenticator.kt (hunk 2 above)
Comment (sschuberth) [reply utils/src/main/kotlin/OrtAuthenticator.kt:109]: Done.

**Comment by sschuberth** [comment]
Merging this despite the unrelated *GoDepTest* failure which was meanwhile fixed independently in https://github.com/oss-review-toolkit/ort/pull/2678.

**Pull request merged**